ec/src/group/wnaf.rs

```diff
@@ -0,0 +1,39 @@
+extern crate alloc;
+use crate::Group;
+use alloc::vec::Vec;
+
+pub fn wnaf_mul<G: Group>(table: &[G], scalar_wnaf: &[i64]) -> G {
+    let mut result = G::zero();
+
+    let mut found_one = false;
+
+    for n in scalar_wnaf.iter().rev() {
+        if found_one {
+            result = result.double();
+        }
+
+        if *n != 0 {
+            found_one = true;
+
+            if *n > 0 {
+                result += &table[(n / 2) as usize];
+            } else {
+                result -= &table[((-n) / 2) as usize];
+            }
+        }
+    }
+
+    result
+}
+
+pub fn wnaf_table<G: Group>(table: &mut Vec<G>, mut base: G, window: usize) {
+    table.truncate(0);
+    table.reserve(1 << (window - 1));
+
+    let dbl = base.double();
+
+    for _ in 0..(1 << (window - 1)) {
```

Comment:
Once this is merged, lets make an issue to add parallelism to this

Reply:
not sure you can do this. Base is mutated sequentially

Reply:
You can, its similar to computing all powers of a generator in parallel. You chunk which multiples are to be computed in each thread, and then just combine them together when building the final window.

Reply:
Fair enough, but typically, w =5 or 6 at most for wNAF

Reply:
Furthermore, for your idea, where it applies, one is probably still better off doubling sequentially to create a table and then running the exponentiation in parallel on the back of that table